### PR TITLE
op-chain-ops/script: turn any Go struct/interface into a precompile with ABI encoding/decoding

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

This lets you turn any Go object into a precompile with automatic ABI handling.

This is a lot like the Geth RPC server, where you can put some public methods on an object, and it turns into JSON-RPC handled routes.

Using reflection, we turn struct fields and methods of an object automatically in a table of functions.
Each function is matched by the 4-byte selector, and automatically does the ABI decoding of inputs, and ABI-encoding of outputs.

This makes it trivially easy to put configs, cheat-functions, etc. into the Forge script environment.

**Tests**

Creates an example precompile object, with embedded struct and some different setters/getters, showing off the automatic ABI functionality.

**Additional context**

Part of a PR stack, to run Forge deployment scripts in Go.

### CODE REVIEW

**Review by tynes — APPROVED** [review]
Nice reflection magic, this looks good to me, the only suggestion is more comments so that other people can more easily maintain but fine with merging as is

File: op-chain-ops/script/precompile.go [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+		return encodeRevert(fmt.Errorf("failed to run %s, ABI: %q, err: %w", fn.goName, fn.abiSignature, err))
+	}
+	return out, nil
+}
+
+var revertSelector = crypto.Keccak256([]byte("Error(string)"))[:4]
+
+func encodeRevert(outErr error) ([]byte, error) {
+	outErrStr := []byte(outErr.Error())
+	out := make([]byte, 0, 4+32*2+len(outErrStr)+32)
```
Comment (axelKingsley) [inline op-chain-ops/script/precompile.go:451]: Could you put a comment on this one? Same with other magic values below, maybe a function-comment would do well.

File: op-chain-ops/script/precompile.go [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+	abiFunctionName := fieldDef.Name
+	if len(abiFunctionName) == 0 {
+		return fmt.Errorf("ABI name of %s must not be empty", fieldDef.Name)
+	}
+	if lo := strings.ToLower(abiFunctionName[:1]); lo != abiFunctionName[:1] {
+		abiFunctionName = lo + abiFunctionName[1:]
+	}
+
+	methodSig := abiFunctionName + "()"
+	byte4Sig := bytes4(methodSig)
```
Comment (axelKingsley) [inline op-chain-ops/script/precompile.go:387]: This seems a little fragile, are there any `fieldDef.Name` that wouldn't work?

**Review by axelKingsley — APPROVED** [review]
A lot of the ABI specifics went over my head, but I was able to look over the reflection aspects. 

There are some string-manipulations that seem fragile, I left a comment on the one that stuck out to me. But for the sake of driving this forward with the Forge Scripts in Go, I think this is good to go on to the next stage. If we see issues when using this, we can come back and improve things.

File: op-chain-ops/script/precompile.go (hunk 1 above)
Comment (protolambda) [reply op-chain-ops/script/precompile.go:451]: It's ABI encoding of a string function argument, following a solidity error signature. Added comments to clarify.

File: op-chain-ops/script/precompile.go (hunk 2 above)
Comment (protolambda) [reply op-chain-ops/script/precompile.go:387]: Added comments to clarify what is happening here. There is some validation of the field definition already, all field names would work (if the name is a sensible ABI signature is up to the user).

**Comment by roninjin10** [comment]
This is so cool!

**Pull request merged**